akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -883,24 +910,126 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
+     * with an empty response entity.
+     *
+     * @param hardDeadline timeout after which all requests and connections shall be forcefully terminated
+     * @return future which completes successfully with a marker object once all connections have been terminated
+     */
+    def terminate(hardDeadline: FiniteDuration): Future[HttpTerminated] = {
+      require(hardDeadline > Duration.Zero, "deadline must be greater than 0, was: " + hardDeadline)
+
+      _whenTerminationSignalIssued.trySuccess(hardDeadline.fromNow)
+      val terminated = unbindAction().flatMap(_ ⇒ terminateAction(hardDeadline))(ExecutionContexts.sameThreadExecutionContext)
+      _whenTerminated.completeWith(terminated)
+      whenTerminated
+    }
+
+    /**
+     * Completes when the [[terminate]] is called and server termination is in progress.
+     * Can be useful to make parts of your application aware that termination has been issued,
+     * and they have [[Deadline]] time remaining to clean-up before the server will forcefully close
+     * existing connections.
+     *
+     * Note that while termination is in progress, no new connections will be accepted (i.e. termination implies prior [[unbind]]).
+     */
+    def whenTerminationSignalIssued: Future[Deadline] =
+      _whenTerminationSignalIssued.future
+
+    /**
+     * This future completes when the termination process, as initiated by an [[terminate]] call has completed.
+     * This means that the server is by then: unbound, and has closed all existing connections.
+     *
+     * This signal can for example be used to safely terminate the underlying ActorSystem.
+     *
+     * Note: This mechanism is currently NOT hooked into the Coordinated Shutdown mechanisms of Akka.
+     *       TODO: This feature request is tracked by: https://github.com/akka/akka-http/issues/1210
+     *
+     * Note that this signal may be used for Coordinated Shutdown to proceed to next steps in the shutdown.
+     * You may also explicitly depend on this future to perform your next shutting down steps.
+     */
+    def whenTerminated: Future[HttpTerminated] =
+      _whenTerminated.future
+
   }
 
+  /** Type used to carry meaningful information when server termination has completed successfully. */
+  @DoNotInherit sealed abstract class HttpTerminated extends akka.http.javadsl.HttpTerminated
+  sealed abstract class HttpServerTerminated extends HttpTerminated
+  object HttpServerTerminated extends HttpServerTerminated
+  sealed abstract class HttpConnectionTerminated extends HttpTerminated
+  object HttpConnectionTerminated extends HttpConnectionTerminated
+
   /**
    * Represents one accepted incoming HTTP connection.
    */
   final case class IncomingConnection(
     localAddress:  InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow:          Flow[HttpResponse, HttpRequest, NotUsed]) {
+    _flow:         Flow[HttpResponse, HttpRequest, ServerTerminator]) { // FIXME may not change mat value I guess
```

Comment:
As long as the `NotUsed` instance is not used... or could it lead to a `ClassCastException` somewhere even though the actual value was ignored?

Reply:
FIXME?

Reply:
Removed fixme... I think it could be an issue if someone did mapMaterialized value on a flow obtained by pattern matching i guess... 

Note that for just accessing the flow I kept the `def flow` below so .flow would still be the same type and not cause issues in mapMaterializedValue.